/* Copyright 2022 The MediaPipe Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

syntax = "proto2";

package mediapipe.tasks.vision.object_detector.proto;

import "mediapipe/framework/calculator.proto";
import "mediapipe/framework/calculator_options.proto";
import "mediapipe/tasks/cc/core/proto/base_options.proto";

option java_package = "com.google.mediapipe.tasks.vision.objectdetector.proto";
option java_outer_classname = "ObjectDetectorOptionsProto";

message ObjectDetectorOptions {
  extend mediapipe.CalculatorOptions {
    optional ObjectDetectorOptions ext = 443442058;
  }
  // Base options for configuring MediaPipe Tasks, such as specifying the TfLite
  // model file with metadata, accelerator options, etc.
  optional core.proto.BaseOptions base_options = 1;

  // The locale to use for display names specified through the TFLite Model
  // Metadata, if any. Defaults to English.
  optional string display_names_locale = 2 [default = "en"];

  // The maximum number of top-scored detection results to return. If < 0, all
  // available results will be returned. If 0, an invalid argument error is
  // returned. Note that models may intrinsically be limited to returning a
  // maximum number of results N: if the provided value here is above N, only N
  // results will be returned.
  optional int32 max_results = 3 [default = -1];

  // Score threshold to override the one provided in the model metadata (if
  // any). Detection results with a score below this value are rejected.
  optional float score_threshold = 4;

  // Optional allowlist of category names. If non-empty, detection results whose
  // category name is not in this set will be filtered out. Duplicate or unknown
  // category names are ignored. Mutually exclusive with category_denylist.
  repeated string category_allowlist = 5;

  // Optional denylist of category names. If non-empty, detection results whose
  // category name is in this set will be filtered out. Duplicate or unknown
  // category names are ignored. Mutually exclusive with category_allowlist.
  repeated string category_denylist = 6;

  // Whether to use multiclass NMS. That is, each category processes
  // non-maximum-suppression separately.
  optional bool multiclass_nms = 7 [default = false];

  // Overlapping threshold for non-maximum-suppression calculator. Only used for
  // models without built-in non-maximum-suppression, i.e., models that don't
  // use the Detection_Postprocess TFLite Op
  optional float min_suppression_threshold = 8 [default = 0.3];
}
